abmarl/sim/gridworld/actor.py

```diff
@@ -147,69 +166,366 @@ def attack_mapping(self, value):
+        else:
+            return []
+
+
+class EncodingBasedAttackActor(AttackActorBaseComponent):
+    """
+    Agents can attack other agents.
+
+    The attacking agent specifies how many attacks it would like to use per available
+    encoding, based on its attack count and the attack mapping. For example, if
+    the agent can attack encodings 1 and 2 and has up to 3 attacks available, then
+    it may specify up to 3 attacks on encoding 1 and up to 3 attack on encoding 2.
+    Agents with those encodings in the surrounding grid are liable to be attacked.
+    """
+    def _assign_space(self, agent):
+        attackable_encodings = self.attack_mapping[agent.encoding]
+        agent.action_space[self.key] = Dict({
+            i: Discrete(agent.attack_count + 1) for i in attackable_encodings
+        })
+        agent.null_action[self.key] = {i: 0 for i in attackable_encodings}
 
-        The processing goes through a series of checks. The attack is possible
-        if there is an attacked agent such that:
+    def _determine_attack(self, agent, attack):
+        """
+        Process the agent's attack.
 
-        1. The attacked agent is active.
-        2. The attacked agent is within range.
-        3. The attacked agent is valid according to the attack_mapping.
+        The agent specifies how many attacks to carry out per encoding. The
+        EncodingBasedAttackActor searches the nearby local grid defined by the
+        agent's attack range for attackable agents, and randomly chooses from that
+        set up to the number of attacks issued for each encoding.
 
-        If the attack is possible, then we determine the success of the attack
-        based on the attacking agent's accuracy. If the attack is successful, then
-        the attacked agent's health is depleted by the attacking agent's strength,
-        possibly resulting in its death.
+        Args:
+            agent: The attacking agent.
+            attack: The number of attacks to perform per each encoding.
+        Returns:
+            List of attacked agents, up to the number of attacks the agent can
+            carry out per encoding per step. The agent's attack count is an upper bound
+            on the attack per encoding, not a total upper bound.
         """
-        def determine_attack(agent):
-            # Generate local grid and an attack mask.
-            local_grid, mask = gu.create_grid_and_mask(
-                agent, self.grid, agent.attack_range, self.agents
+        # Return empty list if no attack is specified.
+        if not any([num_attacks for num_attacks in attack.values()]):
+            return []
+
+        # Generate local grid and an attack mask.
+        local_grid, mask = gu.create_grid_and_mask(
+            agent, self.grid, agent.attack_range, self.agents
+        )
+
+        # Randomly scan the local grid for attackable agents.
+        attackable_agents = {encoding: [] for encoding in attack}
```

Comment:
Additional comments: This attack actor processes the attacking in the same way as the BinaryAttackActor; the only difference is that it does the processing for each encoding. We could have designed this differently to avoid code-duplication, but we left it as is (1) for clarity of code and (2) because we believe our implementation is more performant since we only have to scan the local grid a single time.